FFmpegInterop/H264SampleProvider.cpp

```diff
@@ -18,55 +18,209 @@
+		m_data(data),
+		m_dataSize(dataSize)
 	{
-		// Call base class method that simply write the packet to stream as is
-		hr = MediaSampleProvider::WriteAVPacketToStream(dataWriter, avPacket);
+		// Validate parameters
+		WINRT_ASSERT(m_data != nullptr);
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_dataSize < MIN_SIZE);
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_data[0] != 1);
 	}
 
-	// We have a complete frame
-	return hr;
-}
+	uint8_t AVCConfigParser::GetNaluLengthSize() const noexcept
+	{
+		return m_data[4] & 0x03 + 1;
+	}
 
-HRESULT H264SampleProvider::GetSPSAndPPSBuffer(DataWriter^ dataWriter)
-{
-	HRESULT hr = S_OK;
+	tuple<vector<uint8_t>, vector<uint32_t>> AVCConfigParser::GetSpsPpsData() const
+	{
+		vector<uint8_t> spsPpsData;
+		vector<uint32_t> spsPpsNaluLengths;
+		uint32_t pos{ 5 };
+
+		// Parse SPS NALUs
+		uint8_t spsCount{ static_cast<uint8_t>(m_data[pos++] & 0x1F) };
+		pos += ParseParameterSets(spsCount, m_data + pos, m_dataSize - pos, spsPpsData, spsPpsNaluLengths);
+
+		// Parse PPS NALUs
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, pos >= m_dataSize);
+		uint8_t ppsCount{ m_data[pos++] };
+		ParseParameterSets(ppsCount, m_data + pos, m_dataSize - pos, spsPpsData, spsPpsNaluLengths);
+
+		return { spsPpsData, spsPpsNaluLengths };
+	}
+
+	uint32_t AVCConfigParser::ParseParameterSets(
+		_In_ uint8_t parameterSetCount,
+		_In_reads_(dataSize) const uint8_t* data,
+		_In_ uint32_t dataSize,
+		_Inout_ vector<uint8_t>& spsPpsData,
+		_Inout_ vector<uint32_t>& spsPpsNaluLengths) const
+	{
+		// Reserve estimated space now to minimize reallocations
+		spsPpsNaluLengths.reserve(spsPpsNaluLengths.size() + parameterSetCount);
+
+		// Parse the parameter sets and convert the SPS/PPS data to Annex B format
+		uint32_t pos{ 0 };
+		for (uint8_t i{ 0 }; i < parameterSetCount; i++)
+		{
+			// Get the NALU length
+			uint32_t naluLength{ GetAVCNaluLength(data + pos, dataSize - pos, sizeof(uint16_t)) };
+			pos += sizeof(uint16_t);
 
-	if (!m_pAvCodecCtx || (m_pAvCodecCtx->extradata == nullptr && m_pAvCodecCtx->extradata_size < 8))
+			// Write the NALU start code and SPS/PPS data to the buffer
+			spsPpsData.insert(spsPpsData.end(), begin(NALU_START_CODE), end(NALU_START_CODE));
+			spsPpsData.insert(spsPpsData.end(), data + pos, data + pos + naluLength);
+
+			spsPpsNaluLengths.push_back(sizeof(NALU_START_CODE) + naluLength);
+
+			pos += naluLength;
+		}
+
+		return pos; // Return the number of bytes read
+	}
+
+	bool AVCConfigParser::HasNoFMOASO() const
 	{
-		// The data isn't present
-		hr = E_FAIL;
+		uint32_t pos{ 5 };
+
+		// Scan the SPS NALUs and check if any have a non-constrained baseline
+		bool foundNonConstrainedBaselineSPS{ false };
+		uint8_t spsCount{ static_cast<uint8_t>(m_data[pos++] & 0x1F) };
+
+		for (uint8_t i{ 0 }; i < spsCount; i++)
+		{
+			// Get the SPS NALU length
+			uint32_t naluLength{ GetAVCNaluLength(m_data + pos, m_dataSize - pos, sizeof(uint16_t)) };
```

Comment:
>m_dataSize - pos [](start = 55, length = 16)

The comments from ParseParameterSets apply to this method as well - we'll need a bounds check here and below

Reply:
Resolving as this is also addressed by the bounds check in GetAVCNaluLength